docs/proposals/004-com-support.md

```diff
@@ -0,0 +1,240 @@
+{
+    return stuff.t;
+}
+HRESULT someOperations(IDevice* device, ComPtr<ITexture>& _result)
+{
+    ComPtr<ITexture> t;
+    HRESULT _err = device->createTexture(&t);
+    if(_err < 0) return _err;
+
+    _result = t;
+    return 0;
+}
+```
+
+As a small optimization, an `in` parameter of a `[COM] interface` type translates as a C++ parameter of just the matching pointer type (see `device` above).
+
+Note that the translation of idiomatic `HRESULT` return codes into `throws HRESULT` functions in Slang allows code working with COM interfaces to benefit from the convenience of the Slang error handling model.
+
+
+Detailed Explanation
+--------------------
+
+This is a case where the simple explanation above covers most of the interesting stuff.
+
+There are a lot of semantic checks we'd need/want to implement to make sure `[COM]` interfaces are used correctly:
+
+* Any `interface` that inherits from one or more other `[COM]` interfaces must itself be `[COM]`
+
+* Any concrete type that implements a `[COM]` interface must be a `class`
+
+There are also detailed implementation questions to be answered around the in-memory layout of `class` types that implement `[COM]` interfaces.
+In particular, we might want to be able to optimize for the case of a single-inheritance `class` hierarchy that mirrors a `[COM] interface` hierarchy, since this comes up often for COM-based APIs:
+
+```
+interface IBase { void baseFunc(); }
+interface IDerived : IBase { void derivedFunc(); }
+
+class BaseImpl : IBase { ... }
+class DerivedImpl : BaseImpl, IDerived { ... }
+```
+
+Using a naive translation to C++, the `DerivedImpl` type could end up with *three* different virtual function table (`vtbl`) pointers embedded in it: one for `slang::Object`, one for `IBase`, and one for `IDerived`.
+Clearly the `vtbl`s for `IBase` and `IDerived` could be shared, but C++ `class`es can't easily express this.
+Furthermore, if we are able to tune our strategy for layout, we can set things up so that `[COM] interface`s consume `vtbl` slots starting at index `0` and counting up, while any `virtual` methods in `class`es consume slots starting at index `-1` and counting *down*.
+Using such a layout strategy we can actually allow a type like `DerivedImpl` above to use only a *single* `vtbl` pointer.
+
+Questions
+--------------
+
+### Should we emit COM code that works at the plain C level, or idiomatic C++?
+
+I honestly don't know. Emitting idomatic C++ (and using things like smart pointers) certainly makes the output code easier to understand.
+
+### Can we make this work with more advanced features of Slang interfaces?
+
+Some Slang features don't have perfect analogues. For example, given that `[COM] interface`s can only be conformed to by `class` types, the use of `[mutating]` isn't especially meaningful.
+
+There is no reason why a `[COM] interface` couldn't make use of `static` methods, but there would be no way to call those from C++ without an instance of the interface type.
```

Comment:
You could have a static member of the interface, but that is problematic. You could bundle up all of the static functions into an interface, that you can access via some GetStaticInterface.. like function. That could return an array of function pointers. Having it return an interface is probably reasonable.